RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1481 @@
+                      ConfigParameters const& configParameters) {
+        auto const totalChannels =
+            f01HEDigis.metadata().size() + f5HBDigis.metadata().size() + f3HBDigis.metadata().size();
+        // FIXME: the number of channels in output might change given that some channesl might be filtered out
+
+        // do not run when there are no rechits (e.g. if HCAL is not being read),
+        // but do set the size of the output collection to 0
+        if (totalChannels == 0) {
+          return;
+        }
+
+        // TODO: this can be lifted by implementing a separate kernel
+        // similar to the default one, but properly handling the diff in #sample
+        // or modifying existing one
+        // TODO: assert startingSample = f01nsamples - windowSize to be 0 or 2
+        // assert f01nsamples == f5nsamples
+        // assert f01nsamples == f3nsamples
+        int constexpr windowSize = 8;
+
+        //compute work division
+        uint32_t nchannels_per_block = configParameters.kprep1dChannelsPerBlock;
+        auto const blocks_y = cms::alpakatools::divide_up_by(totalChannels, nchannels_per_block);
+
+        Vec2D const blocks_2d{blocks_y, 1u};  // {y, x} coordiantes
+        Vec2D const threads_2d{nchannels_per_block, windowSize};
+        auto workDivPrep2D = cms::alpakatools::make_workdiv<Acc2D>(blocks_2d, threads_2d);
+
+        //Device buffer for output
+        auto amplitudes = cms::alpakatools::make_device_buffer<float[]>(queue, totalChannels * windowSize);
+        auto noiseTerms = cms::alpakatools::make_device_buffer<float[]>(queue, totalChannels * windowSize);
+        auto electronicNoiseTerms = cms::alpakatools::make_device_buffer<float[]>(queue, totalChannels * windowSize);
+        auto soiSamples = cms::alpakatools::make_device_buffer<int8_t[]>(queue, totalChannels * windowSize);
+
+        alpaka::exec<Acc2D>(queue,
+                            workDivPrep2D,
+                            mahi::kernel_prep1d_sameNumberOfSamples{},
+                            outputGPU,
+                            f01HEDigis,
+                            f5HBDigis,
+                            f3HBDigis,
+                            mahi,
+                            sipmCharacteristics,
+                            recoParamsWithPS,
+                            configParameters.useEffectivePedestals,
+                            configParameters.sipmQTSShift,
+                            configParameters.sipmQNTStoSum,
+                            configParameters.firstSampleShift,
+                            configParameters.ts4Thresh,
+                            reinterpret_cast<float*>(amplitudes.data()),
+                            reinterpret_cast<float*>(noiseTerms.data()),
+                            reinterpret_cast<float*>(electronicNoiseTerms.data()),
+                            reinterpret_cast<float*>(soiSamples.data()),
```

Comment:
Why are these `reinterpret_cast`s needed? The `amplitudes`, `noiseTerms`, and `elecrtronicNoiseTerms` were allocated as `float` arrays, so the type should be correct out of the box. The `soiSamples` was allocated as `int8_t`, so using it via `float*` looks dangerous.

Reply:
The cast for `soiSamples` is a typo. Changed the type to `int8_t` and removed the `reinterpret_cast` together with others.